File: iOS/LNCollectionView/LNScrollView/Pulse/LNScrollViewPulser.m
//
//  LNScrollViewPulseReceiver.m
//  LNCollectionView
//
//  Created by Levison on 27.11.24.
//

#import "LNScrollViewPulser.h"

@interface LNScrollViewPulser ()

@property (nonatomic, assign) BOOL isOpen;

@end

@implementation LNScrollViewPulser

- (instancetype)init
{
    self = [super init];
    if (self) {
        self.mass = 1.f;
        self.isOpen = NO;
    }
    return self;
}

- (LNScrollViewMomentum *)getCurrentMomentum
{
    LNScrollViewMomentum *currentMomentum = [[LNScrollViewMomentum alloc] init];
    currentMomentum.mass = self.mass;
    if (self.delegate && [self.delegate respondsToSelector:@selector(pulserGetVelocity:)]) {
        CGFloat velocity = [self.delegate pulserGetVelocity:self];
        currentMomentum.velocity = velocity;
    } else {
        currentMomentum.velocity = 0.f;
    }
    return currentMomentum;
}

- (void)updateMomentum:(LNScrollViewMomentum *)momentum
{
    if (!self.isOpen) {
        return;
    }
    if (self.delegate && [self.delegate respondsToSelector:@selector(pulser:updateVelocity:)]) {
        [self.delegate pulser:self updateVelocity:momentum.velocity];
    }
}

- (void)open
{
    self.isOpen = YES;
}

- (void)close
{
    self.isOpen = NO;
}

@end

File: Android/app/src/main/java/com/ln/collectionview/scrollview/pulse/LNScrollViewPulser.kt
package com.ln.collectionview.scrollview.pulse

import java.lang.ref.WeakReference

interface LNScrollViewPulserDelegate {
    fun pulserGetVelocity(pulser: LNScrollViewPulser): Float
    fun pulserUpdateVelocity(pulser: LNScrollViewPulser, velocity: Float)
}

class LNScrollViewPulser {

    var mass: Float = 1f
    private var isOpen: Boolean = false

    // Function to get current momentum
    fun getCurrentMomentum(): LNScrollViewMomentum {
        val currentMomentum = LNScrollViewMomentum()
        currentMomentum.mass = this.mass
        if (delegate != null) {
            val velocity = delegate?.get()?.pulserGetVelocity(this) ?: 0f
            currentMomentum.velocity = velocity
        } else {
            currentMomentum.velocity = 0f
        }
        return currentMomentum
    }

    // Function to update the momentum
    fun updateMomentum(momentum: LNScrollViewMomentum) {
        if (!isOpen) {
            return
        }
        delegate?.get()?.pulserUpdateVelocity(this, momentum.velocity)
    }

    fun openStatus(): Boolean {
        return isOpen
    }

    fun open() {
        isOpen = true
    }

    fun close() {
        isOpen = false
    }

    var delegate: WeakReference<LNScrollViewPulserDelegate>? = null
}

